lib/srv/db/mysql/protocol/packet.go

```diff
@@ -108,103 +53,13 @@ func ParsePacket(conn io.Reader) (Packet, error) {
-			return nil, trace.BadParameter("failed to parse ERR packet: %v", packetBytes)
-		}
-		return &Error{packet: packet, message: string(packetBytes[minLen:])}, nil
-
-	case mysql.COM_QUERY:
-		// Be a bit paranoid and make sure the packet is not truncated.
-		if len(packetBytes) < packetHeaderAndTypeSize {
-			return nil, trace.BadParameter("failed to parse COM_QUERY packet: %v", packetBytes)
-		}
-		// 4-byte packet header + 1-byte payload header, then query text.
-		return &Query{packet: packet, query: string(packetBytes[packetHeaderAndTypeSize:])}, nil
-
-	case mysql.COM_QUIT:
-		return &Quit{packet: packet}, nil
-
-	case mysql.COM_CHANGE_USER:
-		if len(packetBytes) < packetHeaderAndTypeSize {
-			return nil, trace.BadParameter("failed to parse COM_CHANGE_USER packet: %v", packetBytes)
-		}
-		// User is the first null-terminated string in the payload:
-		// https://dev.mysql.com/doc/internals/en/com-change-user.html#packet-COM_CHANGE_USER
-		idx := bytes.IndexByte(packetBytes[packetHeaderAndTypeSize:], 0x00)
-		if idx < 0 {
-			return nil, trace.BadParameter("failed to parse COM_CHANGE_USER packet: %v", packetBytes)
-		}
-		return &ChangeUser{packet: packet, user: string(packetBytes[packetHeaderAndTypeSize : packetHeaderAndTypeSize+idx])}, nil
-
-	case mysql.COM_STMT_PREPARE:
-		packet, ok := parseStatementPreparePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_PREPARE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_SEND_LONG_DATA:
-		packet, ok := parseStatementSendLongDataPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_SEND_LONG_DATA packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_EXECUTE:
-		packet, ok := parseStatementExecutePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_EXECUTE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_CLOSE:
-		packet, ok := parseStatementClosePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_CLOSE packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_RESET:
-		packet, ok := parseStatementResetPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_RESET packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case mysql.COM_STMT_FETCH:
-		packet, ok := parseStatementFetchPacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_FETCH packet: %v", packetBytes)
-		}
-		return packet, nil
-
-	case packetTypeStatementBulkExecute:
-		packet, ok := parseStatementBulkExecutePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_BULK_EXECUTE packet: %v", packetBytes)
-		}
-		return packet, nil
+	if int(packetType) < len(packetParsersByType) && packetParsersByType[packetType] != nil {
+		return packetParsersByType[packetType](packetBytes)
```

Comment:
nit: Can we also propagate this call in stack-trace ? 
```suggestion
		packet, err packetParsersByType[packetType](packetBytes)
		if err != nil {
		   return nil, trace.Wrap(err)
		}
		return packet, nil
```